docs/Config.md
# Config

## Top-level keys

|Key|Description|Example|
|---|---|---|
|file_dir|Directory to mount into job containers (absolute path)|/foo/files/|
|cleanup|Delete job containers after completion. Only needed for debugging|True|
|payloads|List of job configs|see below|
|connectors|List of configs for different external connections|see below|
|workers|Number of jobs to run in parallel (default 2). Keep in mind these are Docker containers, which means they carry some overhead - be conservative when using a non-default value. Also used for SMB upload/download in remote mode.|2|
|settings|Override settings at the config-level (alternative to environment variables)|see below|


## Payload config

|Key|Description|Example|
|---|---|---|
|name|Name of job; only used for dependencies|sharpshooter-js|
|image|Docker image name in local cache|sharpshooter|
|cli|Command-line for the container. This will depend on the entrypoint defined in the dockerfile|--help|
|artifacts|List of files to pull from the job after completion. Files will be placed into the file_dir directory|abc.exe|
|dependencies|List of jobs to run before this one. Only needed when the output of one job is required as input for another|sharpshooter-js|
|store|Store the artifact into a variable for future retrieval by @files CLI token|abc-exe|

**Dependencies**

Jobs without dependencies will be put into a batch that is executed first. 
Then, all jobs with dependencies that have been completed in the previous batch are executed next.
This repeats until all jobs are complete. 
Dependencies cannot be cyclical (e.g. job2 depends on job1 and job1 depends on job2).

## Settings

The `settings` key allows for config-level overrides of settings defined in [Settings.md](Settings.md).
This is an alternative method of providing the settings values to environment variables and will override values specified via environment variables.

The value of `settings` should be a map of key-value pairs where each key is one of the values from the "Config Name" column in [Settings.md](Settings.md).

To check the values of these settings for a run, review the log file after execution completes.

Example for overriding the Mythic GET and POST URIs:

```yaml
settings:
  mythic_http_geturi: test
  mythic_http_posturi: test
```


docs/Settings.md
# Global settings

The following environment variables can be used for fine-tuning internal settings for the tool. Source: `pdcd.settings:GlobalSettings`

|Variable Name|Description|Config Name|Default|
|---|---|---|---|
|PDCD_LOGFILE|Log file path|N/A|.pdcd.log|
|PDCD_CFGDIR|Directory that contains shared configuration settings such as connectors file|N/A|~/.pdcd|
|PDCD_CONNECTORS|Path to Connectors file|connectors_file|PDCD_CFGDIR + "/" + "connectors"|
|PDCD_SMB_SHARE|Share name for remote build server SMB server|smb_share_name|pdcd|
|PDCD_SMB_TARGET|SMB port on remote build server|smb_target_port|445|
|PDCD_SMB_BIND|Local port to bind to for SMB port forward when using remote builds|smb_bind_port|<random high port>|
|PDCD_DOCKER_TARGET|Docker daemon port on remote build server|docker_target_port|2375|
|PDCD_DOCKER_BIND|Local port to bind to for Docker port forward when using remote builds|docker_bind_port|<random high port>|
|PDCD_SHELL_LOGGING|Log external commands execute via `utils.shell()`|shell_logging|True|
|PDCD_MYTHIC_INTERVAL|Callback interval for HTTP/S payloads|mythic_callback_interval|15|
|PDCD_MYTHIC_JITTER|Callback jitter percent|mythic_jitter_percent|30|
|PDCD_MYTHIC_HTTP_GETURI|HTTP/S GET URI|mythic_http_geturi|search|
|PDCD_MYTHIC_HTTP_POSTURI|HTTP/S POST URI|mythic_http_posturi|form|
|PDCD_MYTHIC_HTTP_QUERYURI|HTTP/S query URI|mythic_http_queryuri|query|
|PDCD_MYTHIC_HTTP_UA|HTTP/S user-agent|mythic_http_useragent|Mozilla/5.0 (Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko|
|PDCD_MYTHIC_SMB_PIPENAME|Override pipe name used for Mythic SMB payloads|mythic_smb_pipename|TSVNCache-00000000487ca41a|
|PDCD_DOCKER_MEM_LIMIT|Max memory for Docker|docker_mem_limit|2G|
|PDCD_DOCKER_MEMSWAP_LIMIT|Max swap for Docker|docker_memswap_limit|2G|



CHANGELOG.md
***

# Version 1.2.0 - January 2024

- Add `settings` key to config to allow settings configuration in config file

***

# Version 1.1.0 - May 2023

- Updated `MythicClient` to support Mythic v3.0.0 (and only Mythic v3.0.0+)

***

# Version 1.0.2 - March 2023

- Bug fix in log subcommand

***

# Version 1.0.0 - March 2023

- Initial release

***



pdcd/settings.py
from pydantic import BaseSettings, Field
from pathlib import Path
import os

from .utils import get_user_pdcd_cfg_dir, find_free_local_port


def cfg_file(f: str):
    cfg_env = os.getenv("PDCD_CFGDIR", None)
    cfg_dir: Path
    if cfg_env is None:
        cfg_dir = get_user_pdcd_cfg_dir()
    else:
        cfg_dir = Path(cfg_env)
    cfg_f = cfg_dir / f
    return cfg_f


class GlobalSettings(BaseSettings):
    # general settings
    log_file: str = Field(default=".pdcd.log", env="PDCD_LOGFILE")
    connectors_file: Path = Field(default_factory=lambda: cfg_file("connectors"), env="PDCD_CONNECTORS")

    # docker settings
    docker_mem_limit: str = Field(default="2G", env="PDCD_DOCKER_MEM_LIMIT")
    docker_memswap_limit: str = Field(default="2G", env="PDCD_DOCKER_MEMSWAP_LIMIT")

    # mythic connector settings
    mythic_callback_interval: int = Field(default=15, env="PDCD_MYTHIC_INTERVAL")
    mythic_jitter_percent: int = Field(default=30, env="PDCD_MYTHIC_JITTER")
    mythic_http_geturi: str = Field(default="search", env="PDCD_MYTHIC_HTTP_GETURI")
    mythic_http_posturi: str = Field(default="form", env="PDCD_MYTHIC_HTTP_POSTURI")
    mythic_http_queryuri: str = Field(default="query", env="PDCD_MYTHIC_HTTP_QUERYURI")
    mythic_http_useragent: str = Field(
        default="Mozilla/5.0 (Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko", env="PDCD_MYTHIC_HTTP_UA"
    )
    mythic_smb_pipename: str = Field(default="TSVNCache-00000000487ca41a", env="PDCD_MYTHIC_SMB_PIPENAME")

    # remote connector settings
    smb_share_name: str = Field(default="pdcd", env="PDCD_SMB_SHARE")
    docker_target_port: int = Field(default=2375, env="PDCD_DOCKER_TARGET")
    docker_bind_port: int = Field(default_factory=find_free_local_port, env="PDCD_DOCKER_BIND")
    smb_target_port: int = Field(default=445, env="PDCD_SMB_TARGET")
    smb_bind_port: int = Field(default_factory=find_free_local_port, env="PDCD_SMB_BIND")
    shell_logging: bool = Field(default=True, env="PDCD_SHELL_LOGGING")


global_settings = GlobalSettings()
